dali/python/nvidia/dali/_utils/external_source_impl.py

```diff
@@ -247,7 +255,31 @@ def accepted_arg_count(callable):
     else:
         implicit_args = 1
         callable = callable.__func__
-    return callable.__code__.co_argcount - implicit_args
+    signature = inspect.signature(callable)
+    # TODO(klecki): Do we mention that callable is one of the alternatives?
+    error_msg = (
+        "The `source` callable must accept either 0 or 1 positional arguments to indicate "
+        "whether it accepts the batch or sample indexing information."
+    )
+    for p in signature.parameters.values():
+        if p.kind == inspect.Parameter.VAR_POSITIONAL:
+            raise TypeError(
+                error_msg + f" Found var-positional argument `*{p.name}` which is not allowed."
+            )
+        if p.kind == inspect.Parameter.VAR_KEYWORD:
+            raise TypeError(
+                error_msg + f" Found var-keyword argument `**{p.name}` which is not allowed."
+            )
+        if p.kind == inspect.Parameter.KEYWORD_ONLY:
+            raise TypeError(
+                error_msg + f" Found keyword-only argument `{p.name}` which is not allowed."
+            )
+    result = len(inspect.signature(callable).parameters) - implicit_args
```

Comment:
I thought `parameters` skips the implicit param already.

Reply:
Btw. the `inspect.signature(callable)` is already assigned to a signature var, right?

Reply:
```
    class MethodOne:
        def __call__(self, a):
            pass

    import inspect

    print(inspect.signature(MethodOne()))
    print(inspect.signature(MethodOne.__call__))
```
prints
```
(a)
(self, a)
```

We extract `__call__`, so it doesn't hide the self anymore.